File: python/stryker_cxx/payload_contract.py
"""Payload contract helpers shared by schema validation and report projection."""

from __future__ import annotations

from typing import Any

REPORT_SCHEMA_VERSION = "stryker-cxx.report.v1"
MTE_SCHEMA_VERSION = "2.0"
TOOL_VERSION = "0.1.0"

NATIVE_STATUSES = (
    "KILLED",
    "SURVIVED",
    "BUILD_ERROR",
    "CHECK_ERROR",
    "NO_COVERAGE",
    "TIMEOUT",
    "IGNORED",
    "PENDING",
    "RUNTIME_ERROR",
)

MTE_STATUSES = (
    "Killed",
    "Survived",
    "NoCoverage",
    "Timeout",
    "Ignored",
    "Pending",
    "RuntimeError",
)

NATIVE_TO_MTE_STATUS = {
    "KILLED": "Killed",
    "SURVIVED": "Survived",
    "BUILD_ERROR": "NoCoverage",
    "CHECK_ERROR": "RuntimeError",
    "NO_COVERAGE": "NoCoverage",
    "TIMEOUT": "Timeout",
    "IGNORED": "Ignored",
    "PENDING": "Pending",
    "RUNTIME_ERROR": "RuntimeError",
}

_NATIVE_STATUS_SET = frozenset(NATIVE_STATUSES)
_MTE_STATUS_SET = frozenset(MTE_STATUSES)


def is_native_status(status: Any) -> bool:
    return str(status) in _NATIVE_STATUS_SET


def is_mte_status(status: Any) -> bool:
    return str(status) in _MTE_STATUS_SET


def native_to_mte_status(status: Any) -> str:
    return NATIVE_TO_MTE_STATUS.get(str(status).upper(), "RuntimeError")


def extract_mte_payload(payload: Any) -> dict[str, Any]:
    if not isinstance(payload, dict):
        raise ValueError("expected mutation-testing-elements payload object")

    if payload.get("schemaVersion") == MTE_SCHEMA_VERSION and isinstance(payload.get("files"), dict):
        return payload

    nested = payload.get("mutationTestingElements")
    if isinstance(nested, dict) and isinstance(nested.get("files"), dict):
        nested_schema = nested.get("schemaVersion")
        if nested_schema is None:
            return {**nested, "schemaVersion": MTE_SCHEMA_VERSION}
        if nested_schema == MTE_SCHEMA_VERSION:
            return nested

    raise ValueError(f"unexpected schemaVersion; expected '{MTE_SCHEMA_VERSION}'")


def supported_mte_statuses() -> tuple[str, ...]:
    return MTE_STATUSES


def supported_native_statuses() -> tuple[str, ...]:
    return NATIVE_STATUSES

File: src/payload-contract.js
export const MTE_SCHEMA_VERSION = "2.0";

export const MTE_STATUSES = Object.freeze([
  "Killed",
  "Survived",
  "NoCoverage",
  "Timeout",
  "Ignored",
  "Pending",
  "RuntimeError",
]);

export const SUMMARY_STATUS_FIELDS = Object.freeze({
  Killed: "killed",
  Survived: "survived",
  NoCoverage: "compileErrors",
  Timeout: "timeouts",
  Ignored: "ignored",
  Pending: "pending",
  RuntimeError: "runtimeErrors",
});

const MTE_STATUS_SET = new Set(MTE_STATUSES);

function isObject(value) {
  return !!value && typeof value === "object" && !Array.isArray(value);
}

export function supportedMteStatuses() {
  return [...MTE_STATUSES];
}

export function isMteStatus(status) {
  return MTE_STATUS_SET.has(status);
}

export function normalizeMtePayload(value) {
  if (!isObject(value)) {
    throw new Error("expected mutation-testing-elements payload object");
  }

  if (value.schemaVersion === MTE_SCHEMA_VERSION && isObject(value.files)) {
    return value;
  }

  const nested = value.mutationTestingElements;
  if (isObject(nested) && isObject(nested.files)) {
    if (nested.schemaVersion == null) {
      return { ...nested, schemaVersion: MTE_SCHEMA_VERSION };
    }
    if (nested.schemaVersion === MTE_SCHEMA_VERSION) {
      return nested;
    }
  }

  throw new Error(`unexpected schemaVersion; expected '${MTE_SCHEMA_VERSION}'`);
}

export function assertMteReport(value) {
  const payload = normalizeMtePayload(value);

  if (typeof payload.schemaVersion !== "string") {
    throw new Error("expected mutation-testing-elements payload object");
  }

  if (payload.schemaVersion !== MTE_SCHEMA_VERSION) {
    throw new Error(`unexpected schemaVersion; expected '${MTE_SCHEMA_VERSION}'`);
  }

  if (!isObject(payload.files)) {
    throw new Error("missing files map");
  }

  if (typeof payload.language !== "string" || payload.language.length === 0) {
    throw new Error("missing language");
  }

  for (const [file, entry] of Object.entries(payload.files)) {
    if (!file || typeof file !== "string") {
      throw new Error("file map key must be a string");
    }
    if (!isObject(entry)) {
      throw new Error(`missing mutants entry for file ${file}`);
    }
    if (!Array.isArray(entry.mutants)) {
      throw new Error(`file ${file} missing mutants array`);
    }

    for (const mutant of entry.mutants) {
      if (!isObject(mutant)) {
        throw new Error(`invalid mutant record in ${file}`);
      }
      for (const field of ["id", "mutatorName", "original", "replacement", "status", "location"]) {
        if (!(field in mutant)) {
          throw new Error(`mutant ${file} missing ${field}`);
        }
      }
      if (!isMteStatus(mutant.status)) {
        throw new Error(`unknown mutant status ${mutant.status} in ${file}`);
      }
      const { location } = mutant;
      if (!isObject(location) || !location.start || !location.end) {
        throw new Error(`invalid location object for mutant ${mutant.id}`);
      }
    }
  }

  return payload;
}

export function toPlainMutationList(report) {
  const out = [];
  for (const [file, entry] of Object.entries(report.files || {})) {
    for (const mutant of entry.mutants || []) {
      out.push({ file, ...mutant });
    }
  }
  return out;
}

export function summarizeReport(report) {
  const flat = toPlainMutationList(report);
  const counts = flat.reduce(
    (acc, mut) => {
      acc.total += 1;
      const field = SUMMARY_STATUS_FIELDS[mut.status];
      if (field) acc[field] += 1;
      return acc;
    },
    {
      total: 0,
      killed: 0,
      survived: 0,
      compileErrors: 0,
      timeouts: 0,
      ignored: 0,
      pending: 0,
      runtimeErrors: 0,
    },
  );

  const denominator = Math.max(0, counts.total - counts.ignored);
  const score = denominator ? counts.killed / denominator : 1;
  return {
    schemaVersion: report.schemaVersion,
    language: report.language,
    projectRoot: report.projectRoot,
    total: counts.total,
    killed: counts.killed,
    survived: counts.survived,
    compileErrors: counts.compileErrors,
    timedOut: counts.timeouts,
    ignored: counts.ignored,
    pending: counts.pending,
    runtimeErrors: counts.runtimeErrors,
    score,
  };
}

export function survivors(report) {
  return toPlainMutationList(report).filter((mut) => mut.status === "Survived");
}
